### PR TITLE
Support Spill for Distinct Aggregation

### LINKED ISSUE #7454: Add spilling support for aggregations with distinct.
### Description

Currently we don't support spilling if aggregation nodes has aggregations with 'distinct' like this:

SELECT count(distinct c0) FROM tmp GROUP BY c1;

Need to add the support if we see queries with this are breaching memory limits.

### PR DESCRIPTION
**Pull request by supermem613**
Queries with distinct aggregations cannot spill. With this change, they are now capable of doing so.

SetAccumulator accumulates the distinct inputs and will now expose an API showing what the maximum spill data is (maxSpillSize), which then can be used with extractForSpill to extract the serialized data. The clear method allows is then to drop the memory usage. Once it is time to bring the data back from spill, the addFromSpill API is used.

The AddressableNonNullValueList subordinate structure, used for complex types (e.g.: ROW, MAP) also is extended with serialization / deserialization capabilities via the new getSerializedSize, copySerializedTo and appendSerialized methods, which follow the same pattern of getting a size, then getting the data and being able to put it back. Its free method is made to leave the structure in a re-usable state as well.

Tests are added to AggregationsTest for the overall spill with distinct aggregation, both when it is on or off and including the VARCHAR case so get coverage beyond a scalar case.

Additionally, a new test suite is added with SetAccumulatorTest to cover all the serialization / deserialization logic added for SetAccumulator in all its implementation cases, including VARCHAR and ComplexType.

Finally, I also added a test for serialization / deserialization for AddressableNonNullValueList.

Fixes #7454

### CODE REVIEW

**Comment by aditi-pandit** [comment]
Hi @supermem613 : Sorry I should've linked the PR in the issue. https://github.com/facebookincubator/velox/pull/7791 was also something I authored for this. Let's check with Masha and Xiaomeng about the approach they prefer.

File: velox/exec/DistinctAggregations.cpp [hunk 1]
```diff
@@ -103,6 +103,20 @@ class TypedDistinctAggregations : public DistinctAggregations {
...
+      const VectorPtr& input,
+      vector_size_t index) override {
+    auto* arrayVector = input->as<ArrayVector>();
+    auto* elementsVector = arrayVector->elements()->asFlatVector<StringView>();
+
+    const auto size = arrayVector->sizeAt(index);
+    const auto offset = arrayVector->offsetAt(index);
+    auto* accumulator = reinterpret_cast<AccumulatorType*>(group + offset_);
+
+    accumulator->addFromSpill(*elementsVector, allocator_);
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:117]: It is a bit strange that size and offset are not used. Is there a bug?

File: velox/exec/DistinctAggregations.cpp [hunk 2]
```diff
@@ -185,6 +199,47 @@ class TypedDistinctAggregations : public DistinctAggregations {
     return input->template asUnchecked<RowVector>()->children();
   }
 
+  void extractForSpill(folly::Range<char**> groups, VectorPtr& result) const {
+    auto* arrayVector = result->as<ArrayVector>();
+    arrayVector->resize(groups.size());
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:204]: nit: perhaps, introduce temp variable numGroups to avoid repeating groups.size()

File: velox/exec/DistinctAggregations.cpp [hunk 3]
```diff
@@ -185,6 +199,47 @@ class TypedDistinctAggregations : public DistinctAggregations {
...
+        arrayVector->mutableOffsets(groups.size())->asMutable<vector_size_t>();
+    auto* rawSizes =
+        arrayVector->mutableSizes(groups.size())->asMutable<vector_size_t>();
+
+    size_t totalBytes = 0;
+    vector_size_t offset = 0;
+    for (auto i = 0; i < groups.size(); ++i) {
+      auto* accumulator =
+          reinterpret_cast<AccumulatorType*>(groups[i] + offset_);
+      rawSizes[i] = accumulator->maxSpillSize();
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:216]: This code looks a bit strange. I assume SetAccumulator to contains a list of rows (distinct inputs for the aggregate function). Hence, rawSizes[i] would be the number of distinct rows stored in the accumulator. Hence, why the method that return this is called "maxSpillSize()" and not "size()"? And why is the same method used to update totalBytes? Would you clarify?

File: velox/exec/DistinctAggregations.cpp [hunk 4]
```diff
@@ -185,6 +199,47 @@ class TypedDistinctAggregations : public DistinctAggregations {
...
+      rawOffsets[i] = offset;
+      offset += accumulator->maxSpillSize();
+      totalBytes += accumulator->maxSpillSize();
+    }
+
+    auto& elementsVector = arrayVector->elements();
+    elementsVector->resize(totalBytes);
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:223]: It is super strange to create one element per byte... Am I missing something?

File: velox/exec/DistinctAggregations.cpp [hunk 5]
```diff
@@ -185,6 +199,47 @@ class TypedDistinctAggregations : public DistinctAggregations {
...
+
+    auto* flatVector = arrayVector->elements()->asFlatVector<StringView>();
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:225]: use elementsVector here

File: velox/exec/DistinctAggregations.cpp [hunk 6]
```diff
@@ -185,6 +199,47 @@ class TypedDistinctAggregations : public DistinctAggregations {
...
+    flatVector->resize(1);
+    auto* rawBuffer = flatVector->getRawStringBufferWithSpace(totalBytes, true);
```
Comment (mbasmanova) [inline velox/exec/DistinctAggregations.cpp:227]: I'm too confused to continue reading. Will wait for you to answer the questions. Thanks.

**Review by mbasmanova — COMMENTED** [review]
@supermem613 Thank you for working on this functionality. I started reading the code and got a bit confused. Would you help answer my questions below?

File: velox/exec/DistinctAggregations.cpp (hunk 1 above)
Comment (supermem613) [reply velox/exec/DistinctAggregations.cpp:117]: We are writing all the serialized data in a single entry

File: velox/exec/DistinctAggregations.cpp (hunk 3 above)
Comment (supermem613) [reply velox/exec/DistinctAggregations.cpp:216]: maxSpillSize because when we are serializing each group the size is an upper bound in maxSpillSize and we serialize we have the actual count. The difference is that we are using an std::optional on the nullBytesIndex and if it is not set, we just serialize a char-size telling it is not set. If it is set, we serialize the char-size with 1 and the actual value. Pedantic, but was trying to be precise. Alternatively, we can change it to always serialize this std::optional regardless if it is set (say, with a 0 for the actual value) so the sizes are always exact. Would that be preferred here?

File: velox/exec/DistinctAggregations.cpp (hunk 4 above)
Comment (supermem613) [reply velox/exec/DistinctAggregations.cpp:223]: I was trying to express a byte array really - an opaque set of bytes.

File: velox/exec/DistinctAggregations.cpp (hunk 6 above)
Comment (supermem613) [reply velox/exec/DistinctAggregations.cpp:227]: Answered the questions - glad to change as needed to make the intent clearer. Sorry, first time using these FlatVector constructs, so I may be abusing them.

File: velox/exec/DistinctAggregations.cpp (hunk 4 above)
Comment (supermem613) [reply velox/exec/DistinctAggregations.cpp:224]: I think in this case it should be ARRAY(VARBINARY) as it is an opaque set of bytes -per group-.  Let me submit a change addressing this resize call here to hopefully make it all clearer.

**Review by mbasmanova — COMMENTED** [review]
@supermem613 Would you like to chat with @aditi-pandit to compare the two solutions and decide which one to proceed with and who will be driving the work?

**Comment by aditi-pandit** [comment]
> @supermem613 Would you like to chat with @aditi-pandit to compare the two solutions and decide which one to proceed with and who will be driving the work?

@supermem613 : Would be great to have a conversation about the 2 PRs. I don't see you on the Velox slack. We can connect via email as well. Mine is aditi.pandit@ibm.com.

I'm not available next week, but we can talk after 1/22 if that works for you.

**Comment by supermem613** [comment]
> > @supermem613 Would you like to chat with @aditi-pandit to compare the two solutions and decide which one to proceed with and who will be driving the work?
> 
> @supermem613 : Would be great to have a conversation about the 2 PRs. I don't see you on the Velox slack. We can connect via email as well. Mine is [aditi.pandit@ibm.com](mailto:aditi.pandit@ibm.com).
> 
> I'm not available next week, but we can talk after 1/22 if that works for you.

Sure, let's chat. Any chance I an be added to the Slack? I already sent an email to velox@meta.com last Friday, but no luck. Sending another one now.

**Comment by mbasmanova** [comment]
@supermem613 Would you share your e-mail? I'll send you an invite for Slack.

**Comment by supermem613** [comment]
> @supermem613 Would you share your e-mail? I'll send you an invite for Slack.

marcusma@msn.com

(I'll delete this comment once I get the invite)

**Comment by mbasmanova** [comment]
@supermem613 Sent the invite.

**Comment by supermem613** [comment]
> @supermem613 Sent the invite.

@mbasmanova 

Thank you. Unfortunately it takes me to an okta login which then lands me on a facebook page that says "this page isn't available".  The page is a SAMLRequest.

**Comment by supermem613** [comment]
> @supermem613 Let me try again, but I need the email (didn't save it).

@mbasmanova  Thank you. It's marcusma at msn dot com

**Comment by mbasmanova** [comment]
@supermem613 Got it. You can delete the comment.

**Comment by mbasmanova** [comment]
> Unfortunately it takes me to an okta login which then lands me on a facebook page that says "this page isn't available". The page is a SAMLRequest.

I'm not able to add you again. Getting "duplicate ..." error. Is there no non-okta option available for you to login?

**Comment by supermem613** [comment]
> > Unfortunately it takes me to an okta login which then lands me on a facebook page that says "this page isn't available". The page is a SAMLRequest.
> 
> I'm not able to add you again. Getting "duplicate ..." error. Is there no non-okta option available for you to login?

It's either okta or "I have a guest account", in which case I need an email + password

**Comment by mbasmanova** [comment]
@supermem613 We've made some changes on our end. Would you try again?

**Comment by supermem613** [comment]
> @supermem613 We've made some changes on our end. Would you try again?

It worked - thank you very much!

**Comment by supermem613** [comment]
Following up with @aditi-pandit over slack about this PR and 7791

**Comment by aditi-pandit** [comment]
> Following up with @aditi-pandit over slack about this PR and 7791

@mbasmanova : @supermem613 and I connected. We will be going ahead with 7791. I'll split the PR as you suggested. I'm off next week, but should be able to spend some time on it 1/22-1/25 and after 1/27.

**Comment by mbasmanova** [comment]
@aditi-pandit Aditi, thank you for the update.
